File: api/src/scrapers/socrata.ts
// =py scrapers/socrata
import { NICHES } from '../lib/stages';
import { encodeParams, fetchJson } from './base';

export type Niche = (typeof NICHES)[number];

const BASE_URL = 'https://data.cityofchicago.org';
export const DATASET_ID = 'r5kz-chrr'; // Business Licenses dataset
// =py settings.SOCRATA_PAGE_SIZE
export const SOCRATA_PAGE_SIZE = 1000;

// Maps our niche types to Socrata business_activity search terms.
// Terms validated against live dataset 2026-03-14.
export const NICHE_MAPPING: Record<Niche, string[]> = {
  barbershops: ['hair service'],
  nail_salons: ['nail service'],
  beauty_shops: ['hair, nail, and skin care', 'hair service'],
  beauty_supply: ['hair service', 'nail service'],
  tire_shops: ['sale and storage of tires'],
  bars: ['tavern'],
  smoke_shops: ['tobacco'],
  meat_markets: ['butcher'],
  towing: ['tow truck', 'tow storage'],
  lawn_services: ['landscap'],
  mobile_mechanics: ['motor vehicle repair'],
  used_auto_parts: ['junk peddler'],
  septic_services: ['plumb'],
  veterinarians: ['veterinar'],
  security_services: ['security service'],
};

export interface SocrataEnv {
  SOCRATA_APP_TOKEN?: string;
}

export interface SocrataRow {
  account_number?: string;
  site_number?: string;
  latitude?: string;
  longitude?: string;
  legal_name?: string;
  doing_business_as_name?: string;
  address?: string;
  zip_code?: string;
  license_number?: string;
  license_status?: string;
  license_start_date?: string;
  business_activity?: string;
}

export type LicenseStatus = 'active' | 'expired' | 'revoked' | 'unknown';

export interface NormalizedBusiness {
  account_number: string | null;
  site_number: string | null;
  latitude: number | null;
  longitude: number | null;
  name: string;
  address: string;
  zip_code: string;
  phone: null;
  niche: Niche;
  license_number: string | null;
  license_status: LicenseStatus;
  license_issue_date: string | null;
}

// =py search_businesses — pages through SoQL results until a short page, an empty page, or the limit
export async function searchBusinesses(
  env: SocrataEnv,
  zipCode: string,
  niche: Niche,
  limit?: number,
): Promise<SocrataRow[]> {
  let pageSize = SOCRATA_PAGE_SIZE;
  if (limit && limit < pageSize) pageSize = limit;

  const searchTerms = NICHE_MAPPING[niche] ?? [];
  if (searchTerms.length === 0) {
    console.warn('no_niche_mapping', { niche });
    return [];
  }

  const termConditions = searchTerms.map((term) => `upper(business_activity) like upper('%${term}%')`).join(' OR ');
  const whereClause = `zip_code='${zipCode}' AND (${termConditions})`;

  const allResults: SocrataRow[] = [];
  let offset = 0;

  for (;;) {
    const params: Record<string, string> = {
      $where: whereClause,
      $limit: String(pageSize),
      $offset: String(offset),
      $order: 'legal_name ASC',
    };
    if (env.SOCRATA_APP_TOKEN) params.$$app_token = env.SOCRATA_APP_TOKEN;

    const page = await fetchJson<SocrataRow[]>(`${BASE_URL}/resource/${DATASET_ID}.json?${encodeParams(params)}`);

    if (page.length === 0) break;

    allResults.push(...page);
    console.log('socrata_page_fetched', { zip_code: zipCode, niche, count: page.length, total: allResults.length });

    if (page.length < pageSize) break;
    if (limit && allResults.length >= limit) return allResults.slice(0, limit);

    offset += pageSize;
  }

  return allResults;
}

// Socrata types every column as text, latitude included.
function num(value: string | undefined): number | null {
  if (!value) return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

// =py normalize_result, plus the account/site identifiers and coordinates Python leaves on the
// table. The city geocodes 92% of licence rows, so these remove a reason to call Google Places.
export function normalizeResult(raw: SocrataRow, niche: Niche): NormalizedBusiness {
  return {
    account_number: raw.account_number ?? null,
    site_number: raw.site_number ?? null,
    latitude: num(raw.latitude),
    longitude: num(raw.longitude),
    name: raw.doing_business_as_name || raw.legal_name || '',
    address: raw.address ?? '',
    zip_code: raw.zip_code ?? '',
    phone: null, // Not in Socrata data
    niche,
    license_number: raw.license_number ?? null,
    license_status: mapLicenseStatus(raw.license_status),
    license_issue_date: raw.license_start_date ?? null,
  };
}

// =py _map_license_status — dataset legend: AAI = license issued, AAC = cancelled during its term,
// REV = revoked, REA = revocation appealed.
export function mapLicenseStatus(status: string | null | undefined): LicenseStatus {
  if (!status) return 'unknown';
  const lower = status.toLowerCase();
  if (lower.includes('aai') || lower.includes('active')) return 'active';
  if (lower.includes('rev')) return 'revoked';
  return 'expired';
}

// One licence row is one renewal, not one business — a single shop can occupy dozens of rows.
// Collapsing them on the city's own account/site key before enrichment keeps a paid Places lookup
// from being spent on the same storefront repeatedly. The most recent licence wins, so status and
// issue date describe the business today. Rows with no account number fall back to name+zip.
export function dedupeLicenseRows(rows: NormalizedBusiness[]): NormalizedBusiness[] {
  const latest = new Map<string, NormalizedBusiness>();
  for (const row of rows) {
    const key = row.account_number
      ? `${row.account_number}/${row.site_number ?? ''}`
      : `${row.name.toLowerCase()}|${row.zip_code}`;
    const seen = latest.get(key);
    if (!seen || (row.license_issue_date ?? '') > (seen.license_issue_date ?? '')) latest.set(key, row);
  }
  return [...latest.values()];
}

File: src/leadforge/scrapers/socrata.py
import structlog

from leadforge.config import settings
from leadforge.db.models.business import NicheType
from leadforge.scrapers.base import BaseAPIClient

logger = structlog.get_logger()

# Maps our niche types to Socrata business_activity search terms
# Chicago Data Portal business licenses dataset: data.cityofchicago.org resource ID "r5kz-chrr"
# Terms validated against live dataset 2026-03-14
NICHE_MAPPING: dict[NicheType, list[str]] = {
    NicheType.BARBERSHOPS: ["hair service"],
    NicheType.NAIL_SALONS: ["nail service"],
    NicheType.BEAUTY_SHOPS: ["hair, nail, and skin care", "hair service"],
    NicheType.BEAUTY_SUPPLY: ["hair service", "nail service"],
    NicheType.TIRE_SHOPS: ["sale and storage of tires"],
    NicheType.BARS: ["tavern"],
    NicheType.SMOKE_SHOPS: ["tobacco"],
    NicheType.MEAT_MARKETS: ["butcher"],
    NicheType.TOWING: ["tow truck", "tow storage"],
    NicheType.LAWN_SERVICES: ["landscap"],
    NicheType.MOBILE_MECHANICS: ["motor vehicle repair"],
    NicheType.USED_AUTO_PARTS: ["junk peddler"],
    NicheType.SEPTIC_SERVICES: ["plumb"],
    NicheType.VETERINARIANS: ["veterinar"],
    NicheType.SECURITY_SERVICES: ["security service"],
}


class SocrataClient(BaseAPIClient):
    """Client for Chicago Data Portal (Socrata) business licenses."""

    DATASET_ID = "r5kz-chrr"  # Business Licenses dataset

    def __init__(self):
        super().__init__(base_url="https://data.cityofchicago.org")

    async def search_businesses(
        self, zip_code: str, niche: NicheType, limit: int | None = None
    ) -> list[dict]:
        """Search businesses by zip code and niche using SoQL."""
        page_size = settings.SOCRATA_PAGE_SIZE
        if limit and limit < page_size:
            page_size = limit

        search_terms = NICHE_MAPPING.get(niche, [])
        if not search_terms:
            logger.warning("no_niche_mapping", niche=niche.value)
            return []

        # Build WHERE clause: zip_code match AND (term1 OR term2 OR ...)
        term_conditions = " OR ".join(
            f"upper(business_activity) like upper('%{term}%')" for term in search_terms
        )
        where_clause = f"zip_code='{zip_code}' AND ({term_conditions})"

        all_results = []
        offset = 0

        client = await self._get_client()

        while True:
            params = {
                "$where": where_clause,
                "$limit": page_size,
                "$offset": offset,
                "$order": "legal_name ASC",
            }
            if settings.SOCRATA_APP_TOKEN:
                params["$$app_token"] = settings.SOCRATA_APP_TOKEN

            response = await client.get(
                f"/resource/{self.DATASET_ID}.json",
                params=params,
            )
            response.raise_for_status()
            page = response.json()

            if not page:
                break

            all_results.extend(page)
            logger.info(
                "socrata_page_fetched",
                zip_code=zip_code,
                niche=niche.value,
                count=len(page),
                total=len(all_results),
            )

            if len(page) < page_size:
                break
            if limit and len(all_results) >= limit:
                all_results = all_results[:limit]
                break

            offset += page_size

        return all_results

    def normalize_result(self, raw: dict, niche: NicheType) -> dict:
        """Normalize a Socrata result into our Business-compatible dict."""
        return {
            "name": raw.get("doing_business_as_name") or raw.get("legal_name", ""),
            "address": raw.get("address", ""),
            "zip_code": raw.get("zip_code", ""),
            "phone": None,  # Not in Socrata data
            "niche": niche,
            "license_number": raw.get("license_number"),
            "license_status": self._map_license_status(raw.get("license_status")),
            "license_issue_date": raw.get("license_start_date"),
        }

    @staticmethod
    def _map_license_status(status: str | None) -> str:
        # Dataset legend: AAI = license issued, AAC = cancelled during its term,
        # REV = revoked, REA = revocation appealed.
        if not status:
            return "unknown"
        status_lower = status.lower()
        if "aai" in status_lower or "active" in status_lower:
            return "active"
        if "rev" in status_lower:
            return "revoked"
        return "expired"
